browser/src/statechart.js

```diff
@@ -0,0 +1,615 @@
+  entry: assign({ trace: [] }),
+  states: {
+    idle: {
+      on: {
+        mousedown: {
+          // prevents select and paint at the same time
+          cond: (_, event) => !event.shiftKey,
+          target: 'dragging',
+          actions: 'addToTrace'
+        }
+      }
+    },
+    dragging: {
+      on: {
+        mousemove: { actions: 'addToTrace' },
+        mouseup: { actions: 'draw' },
+      }
+    }
+  }
+};
+
+const thresholdState = {
+  initial: 'idle',
+  states: {
+    idle: {
+      on: {
+        mousedown: {
+          target: 'dragging',
+          actions: 'storeClick'
+        }
+      }
+    },
+    dragging: {
+      on: {
+        mouseup: [
+          {
+            cond: 'nonemptyBox',
+            actions: 'threshold'
+          },
+          { target: 'idle' },
+        ]
+      }
+    }
+  },
+  on: {
+    TOGGLERGB: {
+      target: 'paint',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+const floodState = {
+  on: {
+    click: { actions: 'flood' }
+  }
+};
+
+const trimState = {
+  on: {
+    click: { actions: 'trim' }
+  }
+};
+
+const erodeDilateState = {
+  on: {
+    mousedown: {
+      actions: choose([
+        {
+          cond: 'leftMouse',
+          actions: 'erode'
+        },
+        {
+          cond: 'rightMouse',
+          actions: 'dilate'
+        }
+      ])
+    }
+  }
+};
+
+const autofitState = {
+  on: {
+    click: { actions: 'autofit' },
+    TOGGLERGB: {
+      target: 'paint',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+const watershedState = {
+  entry: 'storeClick',
+  initial: 'idle',
+  states: {
+    idle: {
+      on: {
+        click: {
+          cond: 'validSeed',
+          target: 'clicked',
+          actions: 'storeClick'
+        }
+      }
+    },
+    clicked: {
+      on: {
+        click: {
+          cond: 'validSecondSeed',
+          actions: 'watershed'
+        }
+      }
+    }
+  },
+  on: {
+    TOGGLERGB: {
+      target: 'paint',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+/**
+ * Handles which current tool for mouse behavior.
+ */
+const toolbarState = {
+  initial: 'paint',
+  states: {
+    flood: floodState,
+    trim: trimState,
+    erodeDilate: erodeDilateState,
+    autofit: autofitState,
+    paint: paintState,
+    threshold: thresholdState,
+    watershed: watershedState,
+    hist: { type: 'history' }
+  },
+  on: {
+    LOADING: 'loading',
+    'keydown.b': '.paint',
+    'keydown.g': '.flood',
+    'keydown.k': '.trim',
+    'keydown.q': '.erodeDilate',
+    // tools not available in RGB mode
+    'keydown.t': {
+      target: '.threshold',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+    'keydown.m': {
+      target: '.autofit',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+    'keydown.w': {
+      target: '.watershed',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+/**
+ * Handles tool behavior with a loading state.
+ */
+const mouseState = {
```

Comment:
<img width="824" alt="Screen Shot 2021-02-16 at 2 19 05 PM" src="https://user-images.githubusercontent.com/40068392/108137319-7f572e00-7070-11eb-8e2f-d444e7bb3ac1.png">

The mouse state controls what clicks do on the canvas. Its main state is the toolbar state, which can be in the flood, trim, erodeDilate, autofit, paint, threshold, or watershed substates. It also has a loading state so clicks do not cause extra actions while results are loading. 

The first four tools (shown above) are simple states that launch an action everytime we click.
 
![Uploading Screen Shot 2021-02-16 at 2.19.12 PM.png…]()

<img width="484" alt="Screen Shot 2021-02-16 at 2 19 27 PM" src="https://user-images.githubusercontent.com/40068392/108137525-e379f200-7070-11eb-8ceb-251082c57c27.png">

The paint state and threshold state have substates to enable clicking and dragging.

<img width="490" alt="Screen Shot 2021-02-16 at 4 06 51 PM" src="https://user-images.githubusercontent.com/40068392/108137585-00162a00-7071-11eb-830e-a9b821aa398e.png">

The watershed state keeps track of two clicks on the same label.

The threshold, watershed and autofit actions are disabled when viewing an RGB image, so these states transition to the paint state on TOGGLERGB events.